notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+     "output_type": "stream",
+     "text": [
+      "2023-10-26 00:59:58.735487 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 0, 59, 58, 735487)"
+      ]
+     },
+     "execution_count": 9,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# ingest clicks_enhanced including user data\n",
+    "sql='''\n",
+    "REPLACE INTO \"clicks_enhanced\" OVERWRITE ALL\n",
+    "WITH \n",
+    "\"users_ext\" AS \n",
+    "(\n",
+    "  SELECT *\n",
+    "  FROM TABLE(\n",
+    "    EXTERN(\n",
+    "      '{\"type\":\"http\",\"uris\":[\"http://datagen:9999/file/users.json\"]}',\n",
+    "      '{\"type\":\"json\"}'\n",
+    "    )\n",
+    "  ) EXTEND (\"time\" VARCHAR, \"user_id\" VARCHAR, \"first_name\" VARCHAR, \"last_name\" VARCHAR, \"dob\" VARCHAR, \"address_lat\" VARCHAR, \"address_long\" VARCHAR, \"marital_status\" VARCHAR, \"income\" VARCHAR, \"signup_ts\" VARCHAR)\n",
+    "),\n",
+    "\"clicks_ext\" AS \n",
+    "(\n",
+    "  SELECT *\n",
+    "  FROM TABLE(\n",
+    "    EXTERN(\n",
+    "      '{\"type\":\"http\",\"uris\":[\"http://datagen:9999/file/clicks.json\"]}',\n",
+    "      '{\"type\":\"json\"}'\n",
+    "    )\n",
+    "  ) EXTEND (\"time\" VARCHAR, \"user_id\" VARCHAR, \"event_type\" VARCHAR, \"client_ip\" VARCHAR, \"client_device\" VARCHAR, \"client_lang\" VARCHAR, \"client_country\" VARCHAR, \"referrer\" VARCHAR, \"keyword\" VARCHAR, \"product\" VARCHAR)\n",
+    ")\n",
+    "SELECT\n",
+    "  TIME_PARSE(c.\"time\") AS \"__time\",\n",
+    "  c.\"user_id\",\n",
+    "  c.\"event_type\",\n",
+    "  c.\"client_ip\",\n",
+    "  c.\"client_device\",\n",
+    "  c.\"client_lang\",\n",
+    "  c.\"client_country\",\n",
+    "  c.\"referrer\",\n",
+    "  c.\"keyword\",\n",
+    "  c.\"product\",\n",
+    "  u.\"first_name\",\n",
+    "  u.\"last_name\",\n",
+    "  u.\"dob\",\n",
+    "  TIMESTAMPDIFF(YEAR, TIME_PARSE(u.\"dob\"), CURRENT_TIMESTAMP) as age,\n",
+    "  ROUND( TIMESTAMPDIFF(YEAR, TIME_PARSE(u.\"dob\"), CURRENT_TIMESTAMP), -1) as age_group,\n",
+    "  u.\"address_lat\",\n",
+    "  u.\"address_long\",\n",
+    "  u.\"marital_status\",\n",
+    "  u.\"income\",\n",
+    "  u.\"signup_ts\"\n",
+    "FROM \"clicks_ext\" c LEFT OUTER JOIN \"users_ext\" u ON c.\"user_id\"=u.\"user_id\"\n",
+    "PARTITIONED BY ALL\n",
+    "'''\n",
+    "display.run_task(sql)\n",
+    "print_timestamp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "350c4465-1196-4fa4-98d6-5a245d4afc20",
+   "metadata": {},
+   "source": [
+    "### Load Users as a Lookup\n",
+    "Lookups are in-memory key/value tables that are pre-broadcast to all brokers, peons and historicals that belong to the lookup tier. They are faster to join to other tables because they are already in memory where they are needed.\n",
```

Comment:
Leave this to the docs team to describe lookups.

Reply:
The whole point of this notebook is to cover the joins subject from end-to-end, I think that if we piece-meal the knowledge and have them jump around multiple notebooks for a given narrative, they lose their usefulness.

But I'm sensing a difference in philosophy for notebook content.